### DISCUSSION TITLE
Polygon Simplification with Douglas-Peucker

### DISCUSSION THREAD

**Discussion by wipfli** [onthegomap/planetiler#534] [Q&A] [answered]
I played around with polygon simplification in QGIS, was quite nice. I used the Douglas-Peucker method there. Can Planetiler do this as well?

**Comment by msbarry** [comment, marked as answer]
Yes, this is what planetiler uses now. You can configure the simplify tolerance (in tile pixels) on each feature that the profile emits. 

The main gap in planetiler is you can't do Visvalingam-Wyatt simplification. I think the VW variant that uses effective area would produce better results generalizing low zoom features (oceans, glaciers, etc) from OSM data instead of needing to use natural earth.

**Reply by wipfli** [reply]
Ah I see, thanks! So in `processFeature` I can already set the pixel tolerance for polygon simplification?

**Reply by msbarry** [reply]
Yes, you can do:

1. `feature.setPixelTolerance(pixels)` to set the simplify tolerance below the max zoom level you are generating
2. `feature.setPixelToleranceAtMaxZoom(pixels)` to set the simplify tolerance at max zoom (separate because you often want more detail when overzooming)

You can also control the default value for 1 with the `--simplify-tolerance` command line argument (defaults to 0.1) and the default value for 2 with `--simplify-tolerance-at-max-zoom` (defaults to 0.0625, the underlying 4096x4096 grid size of the vector tiles).

**Reply by wipfli** [reply]
Great! As far as I understand, polygon simplification gets applied before the post processing step. In my case, I would like to merge buffered buildings in the post processing and then apply polygon simplification to the merged buildings. Is there a way to do this?

**Reply by wipfli** [reply]
Maybe I can just do something like `DouglasPeuckerSimplifier.simplify(merged, tolerance);` on the merged geometries?

**Reply by wipfli** [reply]
Polygon simplification on the merged geometries seems to work. Douglas Peucker with a tolerance of 100 created this image:

![image](https://user-images.githubusercontent.com/53421382/227898467-33d02265-cdb8-44ff-b7fc-8845641463e9.png)

I guess that is a bit too high a tolerance...

**Reply by msbarry** [reply]
This is the code that does the polygon merging:

https://github.com/onthegomap/planetiler/blob/15d0e4f37cf8dc91aeb297b552742aa203f0f7f4/planetiler-core/src/main/java/com/onthegomap/planetiler/FeatureMerge.java#L240-L290

Notice on line 275 that it goes straight from buffered/unbuffered output to snapAndFix without simplifying first. You could insert a simplify step there before snapAndFix and add an extra argument to this method to control simplification factor.

What you're seeing in the result you shared is that simplifying introduces self-intersections in the polygons so they become invalid and maplibre generates these rendering artifacts. That's why you need to simplify before snapAndFix.

**Reply by wipfli** [reply]
Interesting, thanks! Is it possible that the buffering/unbuffering generates self-intersecting polygons which make Douglas Peucker behave in a strange way? I added a `GeoUtils.fixPolygon` step before simplification now, let's see if that changes something...

**Reply by zhibek** [reply]
@msbarry Just picking up this conversation as I've been experimenting with different line simplification/smoothing strategies.

> The main gap in planetiler is you can't do Visvalingam-Wyatt simplification. I think the VW variant that uses effective area would produce better results generalizing low zoom features (oceans, glaciers, etc) from OSM data instead of needing to use natural earth.

Is this simply stating the current situation or are there conceptual reasons that having VW simplification added to planetiler would be problematic or something to be avoided? I realise the performance may be slightly poorer than DP simplification, but the improved output may be worth it.

Depending on your response to the point above, if I get a chance, I'd be interested in trying to implement VW here. Would using JTS source (https://github.com/locationtech/jts/blob/master/modules/core/src/main/java/org/locationtech/jts/simplify/VWLineSimplifier.java#L27) be a reasonable starting point (planning on making similar adaptations to use `MutableCoordinateSequence` as you did in the `DouglasPeuckerSimplifier` class?

**Reply by msbarry** [reply]
> Is this simply stating the current situation or are there conceptual reasons that having VW simplification added to planetiler would be problematic or something to be avoided? I realise the performance may be slightly poorer than DP simplification, but the improved output may be worth it.

This is just the current situation - it would be great to add VW simplification!

> Depending on your response to the point above, if I get a chance, I'd be interested in trying to implement VW here.

Awesome! The 2 pieces to figure out are 1) how to implement it without creating a performance bottleneck and 2) how to expose an API that switches to use VW simplification instead of DP.

For (1) probably makes sense to start out with a benchmark that compares DP with the built-in JTW VW simplification to see how far off it is.  If it's way off, could implement it directly in planetiler.  The VW algorithm is pretty straightforward (https://www.jasondavies.com/simplify/#pseudocode) - I think the JTS implementation is mostly inefficient because it's missing the min heap so just scans through all the possibilities.  We could use `PriorityQueue` for that or the `ArrayLongMinHeap` in `com.onthegomap.planetiler.collection` package.

For (2) I'm not quite sure the best way to switch to this approach, maybe something like `features.line("layer").setPixelTolerance(1).setSimplifyAlgorithm(VISVALINGAM_WYATT);` but then we'd need a way to convert the pixel tolerance between equivalent values for DP and VW.  A more robust approach would switch to a builder for configuring the simplify algorithms like `features.line("layer").transform(GeometryPipeline.create().simplifyVW().useEffectiveArea().minArea(...))` where multiple geometry transforms could be chained, and possibly change by zoom level but that would be a larger refactor.
